File: sensors/hc-sr04-ultraSonic.py
#!/usr/bin/env python3

# This is an example of reading HC-SR04 Ultrasonic Range Finder
# This version measures from the fall of the Trigger pulse 
#   to the end of the Echo pulse
# Check the timing to be sure it calibrated.

import Adafruit_BBIO.GPIO as GPIO
import time

trigger = 'P9_15'    # Pin to trigger the ultrasonic pulse
echo    = 'P9_17'    # Pin to measure to pulse width related to the distance
ms = 1000            # Trigger period in ms

GPIO.setup(trigger, GPIO.OUT)
GPIO.setup(echo, GPIO.IN)

# Called when falling edte of pulse is detected
def pingEnd(channel):
    global startTime
    # print('Edge detected on channel %s'%channel)
    totalTime = time.perf_counter() - startTime
    print('totalTime = {0:0.2f}'.format(1000*totalTime))
    # startTime = time.perf_counter()

GPIO.add_event_detect(echo, GPIO.FALLING, callback=pingEnd) 

# Pulse trigger high then low and start timing.
while True:             # Do something while waiting for event
    # print('ping')
    GPIO.output(trigger, 1)
    time.sleep(0.005)
    GPIO.output(trigger, 0)
    startTime = time.perf_counter()
    time.sleep(ms/1000)         # Wait and let callback to the work
    
GPIO.cleanup()

File: sensors/hc-sr04-ultraSonic.js
#!/usr/bin/env node

// This is an example of reading HC-SR04 Ultrasonic Range Finder
// This version measures from the fall of the Trigger pulse 
//   to the end of the Echo pulse

const b = require('bonescript');
const util = require("util");

const trigger = 'P9_15',  // Pin to trigger the ultrasonic pulse
    echo    = 'P9_17',  // Pin to measure to pulse width related to the distance
    ms = 2500;           // Trigger period in ms
    
var startTime, pulseTime;
    
b.pinMode(echo,    b.INPUT);
b.pinMode(trigger, b.OUTPUT);

b.attachInterrupt(echo, pingEnd, b.FALLING);
 
b.digitalWrite(trigger, 1);     // Unit triggers on a falling edge.
                                // Set trigger to high so we call pull it low later

// Pull the trigger low at a regular interval.
setInterval(ping, ms);

// Pull trigger low and start timing.
function ping() {
    console.log('ping');
    b.digitalWrite(trigger, 0);
    startTime = process.hrtime();
}

// Compute the total time and get ready to trigger again.
function pingEnd(x) {
    // console.log("x: ", util.inspect(x));
    if(x.attached) {
        console.log("Interrupt handler attached: " + x.pin.key);
        return;
    }
    if(startTime) {
        pulseTime = process.hrtime(startTime);
        b.digitalWrite(trigger, 1);
        console.log('pulseTime = ' + (pulseTime[1]/1000000-0.8).toFixed(3));
    }
}
